openedx_tagging/core/tagging/models/base.py

```diff
@@ -395,17 +400,26 @@ def tag_object(
             )
 
         ObjectTagClass = self.object_tag_class
-        current_tags = {
-            tag.tag_ref: tag
-            for tag in ObjectTagClass.objects.filter(
+        current_tags = list(
+            ObjectTagClass.objects.filter(
                 taxonomy=self,
                 object_id=object_id,
             )
-        }
+        )
         updated_tags = []
         for tag_ref in tags:
-            if tag_ref in current_tags:
-                object_tag = current_tags.pop(tag_ref)
+            # Find the existing object_tag index, if any
+            object_tag_index = next(
+                (
+                    i
+                    for i, object_tag in enumerate(current_tags)
+                    if object_tag.tag_ref == tag_ref or object_tag.value == tag_ref
+                ),
+                -1,
+            )
+
+            if object_tag_index >= 0:
+                object_tag = current_tags.pop(object_tag_index)
```

Comment:
I'm using a list here to check the `tag_ref` and also the `value`.
I changed this because the following code was throwing `IntegrityError` on save, violating UNIQUE index.


```python
    def test_tag_object_same_value(self):
        # Tag the object with the same value twice
        tagging_api.tag_object(
            self.taxonomy,
            ["Eubacteria"],
            "biology101",
        )
        tagging_api.tag_object(
            self.taxonomy,
            ["Eubacteria"],
            "biology101",
        )
```

Let me know if this is a problem.

Reply:
I want to understand this change? Why is the `IntegrityError` occurring? From what I see, it is because the value of the tag is being used in closed taxonomies when it is necessary to use the id of the tag. I think that verification should not be done here. It should be done on the view. Because these changes I feel complicate this function.

I think you should revert the change you [made here](https://github.com/openedx/modular-learning/issues/82#issuecomment-1681226213):

> Changed the input from  tags = [{"value": "Interesting stuff"}, {"value": "will surely"}, {"value": "go here"}] to simply tags = ["Interesting stuff", "will surely", "go here"]. The tag_object handle ids and values transparently, so I didn't see a need to treat this in the view explicitly. Let me know if I'm wrong!

In the view it is good to know if it is an id or a value to do the corresponding verifications

Reply:
> From what I see, it is because the value of the tag is being used in closed taxonomies when it is necessary to use the id of the tag. I think that verification should not be done here.
Exactly!

Should I find the Tag from the value and pass its id in this case?

The first `tag_object` call works, but it shouldn't, right? This raises IntegrityError on the second pass:
```python
 def test_tag_object_same_value(self):
        taxonomy = Taxonomy.objects.get(pk=-1)
        # Tag the object with the same value twice
        tagging_api.tag_object(
            taxonomy,
            ["Portuguese"],
            "biology101",
        )
        tagging_api.tag_object(
            taxonomy,
            ["Portuguese"],
            "biology101",
        )
```python

Should we always do this?
```python
 def test_tag_object_same_value(self):
        taxonomy = Taxonomy.objects.get(pk=-1)
        portuguese_tag = Tag.objects.get(value="Portuguese")
        # Tag the object with the same value twice
        tagging_api.tag_object(
            taxonomy,
            [portuguese_tag.id],
            "biology101",
        )
        tagging_api.tag_object(
            taxonomy,
            [portuguese_tag.id],
            "biology101",
        )
```

Reply:
I will take a step back here: I really liked the way the code below just works, and resync does its magic (well done!).
```python
tagging_api.tag_object(
            taxonomy,
            ["Portuguese"],
            "biology101",
        )
```

If you think this is not desirable and can lead to side effects (that I do not know enough to evaluate) I will revert and also find a way to raise an error on this call.

Also, if this code is ok but the find by id and value in the tag_object function let the code unreadable, I can work to try to make it better (maybe creating a function to wrap the check?)

Reply:
Hi @rpenido!

> Also, if this code is ok but the find by id and value in the tag_object function let the code unreadable, I can work to try to make it better (maybe creating a function to wrap the check?)

Yes, it works for me.

Also you need to update the docstring to allow values for closed taxonomies

https://github.com/open-craft/openedx-learning/blob/b1088962e6932691e972c7eee127a2536093f916/openedx_tagging/core/tagging/models/base.py#L384